616 print image process list in nice formatted way

Bajger opened the pull request:
fixes #616

Bajger committed:
Refactored unit tests of image create cmd relying on network, now using stub class to by-pass network use

Bajger committed:
Renamed method for finding template remote url artifact

Bajger committed:
Image finder unit tests do not rely on network communication.

Bajger committed:
Merge branch 'pharo-project:feature/cmd-line' into feature/cmd-line

Bajger committed:
Added support of pretty print of image process list
- Introduced new PhLImageProcess to store data about image process
- fixes #616

Bajger committed:
fixed protocol name

Bajger commented:
Will add unit tests in additional commit

Bajger committed:
Added maximum width of table to print information (35 chars)
Added shortening of string to given size

Bajger committed:
add unit test to verify column width limit, test to shorten the string, fixed protocol name to tests

Bajger committed:
increased column width limit to 50

Bajger committed:
added unit test for image process and test for parsing information from processList command

demarey commented on src/PharoLauncher-CLI/PhLImageProcessListCliCommand.class.st:
```diff
@@ -20,29 +26,46 @@ PhLImageProcessListCliCommand class >> launcherCmdDescription [
-		nextPutAll: self executeOSShellCommand.
-	^ self context exitSuccess 
+	^ (shellOutput substrings: OSPlatform current lineEnding)  collect: [:line |
+			PhLImageProcess newFrom: (line splitOn: '-|-')
```
'-|-' should be factored into a method and all occurences replaced with a call to it

Bajger replied:
Done!

demarey commented on src/PharoLauncher-CLI-Tests/ConsoleListFormatterTest.class.st:
```diff
@@ -294,3 +299,16 @@ ConsoleListFormatterTest >> testPrintRowValuesBy [
+	longest := 'some long string exceeding column width limit'.
+	size := formatter getMaxPrintStringWidthFor: {longer. longest. short.} printBlock: [ :aStr | aStr asString ].
+	
+	"three is the longest word and has a size of 5"
```
I don't understand this comment

Bajger replied:
Added comment to explain in better way.

demarey commented on src/PharoLauncher-CLI-Tests/PhLImageProcessListCliCommandTest.class.st:
```diff
@@ -18,7 +18,28 @@ PhLImageProcessListCliCommandTest >> testExecute [
+			nextPutAll: OSPlatform current lineEnding
+		].
+	resultArray := processListCmd imageProcesListFrom: aString.
+	self assert: resultArray size equals: 3.
```
What are we testing here?

Bajger replied:
Added improved comment in unit test.

demarey commented on src/PharoLauncher-CLI-Tests/StringTest.extension.st:
```diff
@@ -0,0 +1,12 @@
+{ #category : #'*PharoLauncher-CLI-Tests' }
+StringTest >> testShortenTo [
+
+	self assert: ('ABCDEF' shortenTo: 0) equals: ' ... '.
```
' ... ' should be factored into a method also

Bajger replied:
Moved to separate method.

demarey commented on src/PharoLauncher-Core/PhLImageProcess.class.st:
```diff
@@ -0,0 +1,100 @@
+}
+
+{ #category : #private }
+PhLImageProcess class >> listPrintAttributeBlocks [
```
Is this used?

Bajger replied:
Yes, it is used by PhLCliCommand>>listPrintAttributeBlocks that is used in console list formatter:
Line:`self modelClass listPrintAttributeBlocks`,  where modelClass is implemented by PhLImageProcessListCommand and  returns PhLImageProcess.

Bajger committed:
Moved separator to the distinct method in process list command.

Bajger committed:
Improved console list formatter test comments and fixed unit test

Bajger committed:
Improved comment of unit test that instantiates image processes from string lines of OS command  output

Bajger committed:
Factored short form of string to separate method. Adopted unit test.

Bajger commented:
@demarey Hi Christophe! Finally had chance to provide needed changes. Please take a look!

Bajger commented:
Found the issue in parsing line with process information. I didn't test it with pharo processes that run with --headless parameter (or other vm or image params). Therefore parsing is shifted.

hernanmd committed:
Update README.md

Update application start with Spec2 version `PharoLauncherApplication` in Develop the Pharo Launcher section

Bajger committed:
Added sed command to process arbitrary amount of flags or arguments

Bajger committed:
Added debug information to see details of parsed information of OS command.

Bajger committed:
More debug unformatted info about running Pharo processes

Bajger committed:
Fixed array in debug args

Bajger committed:
added debug echo in test result

Bajger committed:
Changed pgrep condition to filter just pharo vm processes

Bajger committed:
Chamges in parsing of line with process (imgPath and vmPath)

demarey committed:
fix template listing in CLI mode

demarey committed:
fix PR artifacts url

demarey committed:
fix wrong VM path in linux packaging

fix #642

demarey committed:
tar linux artifact to avoid executable permission loss

fixes #641

demarey committed:
move PhlErrorDialogPresenter in PharoLancher-Spec2 package

demarey committed:
remove obsolete documentation

Bajger committed:
Added exception handling for creating directory in package command

Bajger committed:
Added version flag to launcher command.

demarey committed:
remove useless method

demarey committed:
launch image after its creation.
Fixes  #648

demarey committed:
Merge branch 'dev' of https://github.com/pharo-project/pharo-launcher into dev

demarey committed:
remove dead code

demarey committed:
Add a 'no-launch' flag to creat image commands.
Related to #648

demarey committed:
do not launch image when creating an image in test setup

demarey committed:
Allow to define a short name on templates and create an image from this template short name.

Fices #647

Bajger committed:
Improved regex validation of GH repository names. (#649)

* Improved regex validation of version in GH repository name.

* Simplified version or branch name regex.

* Updated test for validation of GH repository names.

Bajger committed:
Processing image, vm and pid arguments from line of pgrep command

demarey committed:
add documentation build workflow

demarey committed:
rename docs into doc

Bajger committed:
Fixed obtaining of datetime in ISO format.

Bajger committed:
Fixed parsing of image path, removed debugging log.

demarey committed:
run pharo-launcher CLI without preferences

demarey committed:
Allow fast creation of dev images.
Fixes #656

demarey committed:
rename name flag of list image command to nameFilter for better meaning guess

demarey committed:
update doc

demarey committed:
sort images by last modification (CLI).
Fixes #653

demarey committed:
better test assertion

demarey committed:
archive generated doc if not deployed

demarey committed:
allow deployment of doc to gh pages

demarey committed:
make test non flaky

demarey committed:
Merge pull request #632 from hernanmd/dev

Update README.md

demarey committed:
reorganize commands related to processes.
Fixes #655

demarey committed:
fixes #661

demarey committed:
fix doc of process commands

demarey committed:
Allow to use dev or stable as default image.
Fixes #664

Bajger committed:
Removed case sensitivity in Pharo process identification, added OS Platform specific command arguments

Bajger commented:
I need to resolve OS args passing (on MacOS), here is the situation, when no Pharo process (except launcher) is running: 
```
dbajger@Davids-MacBook-Pro P11-launcher-dev % "/Users/dbajger/Documents/Pharo/vms/110-x64/Pharo.app/Contents/MacOS/Pharo" --headless "/Users/dbajger/Documents/Pharo/images/P11-launcher-dev-proc-list/P11-launcher-dev-proc-list.image" clap launcher image processList
PhLProcessCommandError: OS process command: '/bin/zsh -c pgrep -ai -f 'pharo.*\.image' | xargs ps -o pid,args= -p | grep -v -e 'PID' | grep -v '.bash'| grep -v 'export LD_LIBRARY_PATH' | grep -v 49997'  exited with: normal termination with status 1.
Stderr contents: "".
```
but if I start some pharo image, it works: 
```
dbajger@Davids-MacBook-Pro P11-launcher-dev % "/Users/dbajger/Documents/Pharo/vms/110-x64/Pharo.app/Contents/MacOS/Pharo" --headless "/Users/dbajger/Documents/Pharo/images/P11-launcher-dev-proc-list/P11-launcher-dev-proc-list.image" clap launcher image processList
#  PID   Image name VM path                                            Image path                                         Date/Time started   
-- ----- ---------- -------------------------------------------------- -------------------------------------------------- --------------------
1  50017 P11-OTE    /Users/dbajger/Documen ... pp/Contents/MacOS/Pharo /Users/dbajger/Documen ... s/P11-OTE/P11-OTE.image 2024-01-31 08:49:07
```
Problem is in args after `/bin/zsh -c`, if no process is running then it looks like this:
```
/bin/zsh -c pgrep -ai -f 'pharo.*\.image' | xargs ps -o pid,args= -p | grep -v -e 'PID' | grep -v '.bash'| grep -v 'export LD_LIBRARY_PATH' | grep -v 49997
usage: pgrep [-Lfilnoqvx] [-d delim] [-F pidfile] [-G gid]
             [-P ppid] [-U uid] [-g pgrp] [-t tty] [-u euid]
             pattern ...
```
but if I put it in double qoutes (on terminal) it works:
```
/bin/zsh -c "pgrep -ai -f 'pharo.*\.image' | xargs ps -o pid,args= -p | grep -v -e 'PID' | grep -v '.bash'| grep -v 'export LD_LIBRARY_PATH' | grep -v 49997"

dbajger@Davids-MacBook-Pro P11-launcher-dev %
```
**So I tried to to put each arg to be surrounded by double quoutes:** 
```
osShellArgArray

	^ Array with: self processListCmdArgs surroundedByDoubleQuotes 
```
But it resulted in different error:
```
dbajger@Davids-MacBook-Pro P11-launcher-dev % "/Users/dbajger/Documents/Pharo/vms/110-x64/Pharo.app/Contents/MacOS/Pharo" --headless "/Users/dbajger/Documents/Pharo/images/P11-launcher-dev-proc-list/P11-launcher-dev-proc-list.image" clap launcher image processList
PhLProcessCommandError: OS process command: '/bin/zsh -c "pgrep -ai -f 'pharo.*\.image' | xargs ps -o pid,args= -p | grep -v -e 'PID' | grep -v '.bash'| grep -v 'export LD_LIBRARY_PATH' | grep -v 50592"'  exited with: normal termination with status 127.
Stderr contents: "zsh:1: command not found: pgrep -ai -f 'pharo.*\.image' | xargs ps -o pid,args= -p | grep -v -e 'PID' | grep -v '.bash'| grep -v 'export LD_LIBRARY_PATH' | grep -v 50592
".
```
@demarey Do you have idea, how exactly should be args qouted to OSProcess command? I pass pgrep with all stuff after as one argument (as seen in `osShellArgArray` method).

demarey commented:
Hi David
From the system point of view, the command you are running is `/bin/zsh` and it has 2 arguments: `-c` and a string with all the other parts of the command. Inside the string, you have another command that will be run by `zsh`. It also has to be a valid command. The biggest issue is when you have spaces in arguments. To avoid problems, you can add single or double quotes around all arguments. To summarize, what should be done is:
- run zsh command with a string protected with single (or double) quotes
- protect arguments in the  command string that will be executed bu the shell with double quotes (could be single but not the same used for the zsh command)

Bajger commented:
> run zsh command with a string protected with single (or double) quotes
> protect arguments in the command string that will be executed bu the shell with double quotes (could be single but not the same used for the zsh command)

This is what I've thought of. I've surrounded pgrep and all its args as one argument surrounded by double qoutes. So why I have and error like this? 
```
launcher image processList
PhLProcessCommandError: OS process command: '/bin/zsh -c "pgrep -ai -f 'pharo.*\.image' | xargs ps -o pid,args= -p | grep -v -e 'PID' | grep -v '.bash'| grep -v 'export LD_LIBRARY_PATH' | grep -v 50592"'  exited with: normal termination with status 127.
Stderr contents: "zsh:1: command not found: pgrep -ai -f 'pharo.*\.image' | xargs ps -o pid,args= -p | grep -v -e 'PID' | grep -v '.bash'| grep -v 'export LD_LIBRARY_PATH' | grep -v 50592
".
```
If I put exactly the same string in the terminal: `/bin/zsh -c "pgrep -ai -f 'pharo.*\.image' | xargs ps -o pid,args= -p | grep -v -e 'PID' | grep -v '.bash'| grep -v 'export LD_LIBRARY_PATH' | grep -v 50592"`
It works. 

What looks weird that `command not found`error output prints command with ending quote, but not with starting quote. Maybe some newline character could be there?

demarey commented:
Maybe it has something to do with parameter expansion. In PL, the launch command does exactly what I explained below. Here is an example;
```bash
bash -c 'cd "/Users/me/Documents/Pharo/images/Pharo 11.0" && "/Users/me/Documents/Pharo/vms/110-x64/Pharo.app/Contents/MacOS/Pharo" "/Users/me/Documents/Pharo/images/Pharo 11.0/Pharo 11.0.image"'
```

demarey committed:
enhance temp var names

demarey committed:
Mark default value in  template and category listing. Fixes #663

Bajger committed:
Trimming OS shell command output to prevent further processing errors.

Bajger committed:
Removing whitespace in process list method.

Bajger committed:
Removed debugging printing in shell tests.

Commit:
Update regex filter pattern to match href with Pharo-12 optionally followed by a dot and minor version numbers

demarey committed:
test new regex for Pharo 12

Bajger committed:
Changed single quotes to double quotes in OS shell command, adopted exit code of pgrep.

hernanmd committed:
Merge pull request #1 from pharo-project/enh/pr-670

test new regex for Pharo 12

hernanmd committed:
Update sources.list

Co-authored-by: demarey <christophe.demarey@inria.fr>

demarey committed:
Merge pull request #670 from hernanmd/dev

Fix for including the builds after build 1258

Bajger committed:
More robust image name resolution. Added test for image name.

Bajger commented:
@demarey Please take a look. Process list identification should be robust enough now, I hope I found all cornercases for pgrep. Should be working both on Linux and MacOS (tested).
- I will squash commit and merge to feature/cmd-line
- then cherrypick to dev branch and move fiunctionality to PhLProcessListCliCommand (via new PR)?  

Note: There was a funny problem which I could not identify, since pgrep returns exitCode 1 in case of no process match. This resulted in exception in OS Shell command execution in Pharo - since exitSuccess is tested there. Lessons learned (read man pages) :-)

demarey committed:
Merge branch 'dev' into 616-Print-image-process-list-in-nice-formatted-way

demarey commented:
I did a local merge with dev branch and running tests.
I see a problem with date convertion:
```
OS process command: '/bin/zsh -c ps -o lstart= -p "1974" | xargs -I{} date -jf "%a %b %e %T %Y" "{}" '+%Y-%m-%d %H:%M:%S''  exited with: normal termination with status 1.
Stderr contents: "Failed conversion of ``Lun 29 jan 12:34:40 2024'' using format ``%a %b %e %T %Y''
date: illegal time format
usage: date [-jnRu] [-I[date|hours|minutes|seconds]] [-f input_fmt]
            [-r filename|seconds] [-v[+|-]val[y|m|w|d|H|M|S]]
            [[[[mm]dd]HH]MM[[cc]yy][.SS] | new_date] [+output_fmt]
".
```

demarey committed:
merge PhLImageProcessListCliCommand into PhLProcessListCliCommand

Bajger commented:
> `Lun 29 jan 12:34:40 2024`  
This is date in French localization? I will test with it.

demarey commented:
> > `Lun 29 jan 12:34:40 2024`
> > This is date in French localization? I will test with it.

Yes. And maybe there could be other issues with otherlocalizations. Maybe there is a way to use an ISO format ?

Bajger committed:
Use C locale to standart English formatting when getting date.

Bajger commented:
@demarey: Please try now.

demarey commented:
> @demarey: Please try now.

Not yet

```
OS process command: '/bin/zsh -c LANG=C ps -o lstart= -p "1974" | xargs -I{} date -jf "%a %b %e %T %Y" "{}" '+%Y-%m-%d %H:%M:%S''  exited with: normal termination with status 1.
Stderr contents: "Failed conversion of ``Mon Jan 29 12:34:40 2024'' using format ``%a %b %e %T %Y''
date: illegal time format
usage: date [-jnRu] [-I[date|hours|minutes|seconds]] [-f input_fmt]
            [-r filename|seconds] [-v[+|-]val[y|m|w|d|H|M|S]]
            [[[[mm]dd]HH]MM[[cc]yy][.SS] | new_date] [+output_fmt]
".
```

Bajger commented:
This is strange, if I evaluate on my MacOS:
```
dbajger@Davids-MacBook-Pro P11-launcher-dev % LANG=C ps -o lstart= -p 50442 | xargs -I{} date -jf "%a %b %e %T %Y" "{}" '+%Y-%m-%d %H:%M:%S' 
2024-01-31 09:07:55
```
So on my computer works. I don't know. Date formatting to ISO is really clumsy.

demarey commented:
after investigation the command that works on my mac is:
` date -j -u -f "%a %d %b  %H:%M:%S %Y" '+%Y-%m-%d %H:%M:%S' "Lun 29 jan 12:26:05 2024"`

demarey committed:
refactor PhLImageProcess creation and fix date conversion (from ps) on mac os

Bajger committed:
Replaced input time args in processing start datetime of process.

Bajger commented:
@demarey Tested on my Mac with adopted time attributes (https://github.com/pharo-project/pharo-launcher/pull/617/commits/571fa76afdd7c6bdf14c38a6b69494c1505be832) . Please let me know!

Bajger committed:
Increased table column width limit to 75.

demarey commented:
@Bajger Did you see my PR on this topic on your repository?

Bajger commented:
> @Bajger Did you see my PR on this topic on your repository?

Hi @demarey ! Sorry I didn't recognize that. To resolve conflicts, I need to checkout my branch or https://github.com/pharo-project/pharo-launcher/tree/enh/PR616 ?


<blockquote><img src="https://opengraph.githubassets.com/5ec7b30c102d0a0a24f4475f6e3837842cfd97df8a702fdfffe0892dff4ec105/pharo-project/pharo-launcher" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/pharo-project/pharo-launcher">GitHub - pharo-project/pharo-launcher at enh/PR616</a></strong></div><div>Lets you manage your pharo images and download new ones - GitHub - pharo-project/pharo-launcher at enh/PR616</div></blockquote>

demarey committed:
Merge remote-tracking branch 'Bajger/616-Print-image-process-list-in-nice-formatted-way' into enh/PR616

demarey commented:
> Hi @demarey ! Sorry I didn't recognize that. To resolve conflicts, I need to checkout my branch or https://github.com/pharo-project/pharo-launcher/tree/enh/PR616 ?

I just diid the merge. The PR is up to date 


<blockquote><img src="https://opengraph.githubassets.com/5ec7b30c102d0a0a24f4475f6e3837842cfd97df8a702fdfffe0892dff4ec105/pharo-project/pharo-launcher" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/pharo-project/pharo-launcher">GitHub - pharo-project/pharo-launcher at enh/PR616</a></strong></div><div>Lets you manage your pharo images and download new ones - GitHub - pharo-project/pharo-launcher at enh/PR616</div></blockquote>

Bajger committed:
Merge pull request #1 from pharo-project/enh/PR616

refactor PhLImageProcess creation and fix date conversion (from ps) on mac os
